module-system/sov-modules-api/src/gas.rs

```diff
@@ -0,0 +1,160 @@
+        type Context;
+        type GasConfig: GasConfig;
+
+        fn charge_gas(
+            ws: &mut WorkingSet<Self::Context>,
+            key: &str,
+            price: &<<Self::GasConfig as GasConfig>::GasUnit as GasUnit>::Price,
+        ) -> Result<()> {
+            Self::GasConfig::get_gas_unit(key)
+                .ok_or(anyhow::anyhow!("No gas unit found"))
+                .map(|gas_unit| gas_unit.value(price))
+                .and_then(|gas| ws.charge_gas(gas))
+        }
+    }
+}
+
+pub mod foo {
+    use anyhow::Result;
+
+    use super::sov_modules_api::{GasConfig, GasUnit, Module, WorkingSet};
+
+    // Here we declare a custom gas unit
+
+    pub struct FooGasUnit {
+        pub a: u64,
+        pub b: u64,
+        pub c: u64,
+    }
+
+    impl GasUnit for FooGasUnit {
+        type Price = String;
+
+        fn value(&self, price: &Self::Price) -> u64 {
+            let reduced = price.chars().map(|c| c as u64).sum();
+            self.a
+                .saturating_add(self.b)
+                .saturating_add(self.c)
+                .saturating_mul(reduced)
+        }
+    }
+
+    // The code below will be generated by a macro that will look like
+    // gas_config!(FooGasConfig, FooGasUnit);
+
+    pub struct FooGasConfig {
+        pub keys: [&'static str; 2],
+        pub values: [FooGasUnit; 2],
+    }
+
+    impl GasConfig for FooGasConfig {
+        const MANIFEST: Self = {
+            // read manifest file
+            Self {
+                keys: ["complex_math_operation", "some_other_operation"],
+                values: [
+                    FooGasUnit { a: 1, b: 2, c: 3 },
+                    FooGasUnit { a: 4, b: 5, c: 6 },
+                ],
+            }
+        };
+
+        type GasUnit = FooGasUnit;
+
+        fn get_gas_unit(key: &str) -> Option<&Self::GasUnit> {
+            // this O(n) is an implementation detail and can be freely optimized in the future for
+            // something more efficient; however, we are not expected to have a big number of
+            // attributes for the gas config, so we might not need additional complexity here
+            Self::MANIFEST
+                .keys
+                .iter()
+                .position(|&k| k == key)
+                .and_then(|p| Self::MANIFEST.values.get(p))
+        }
+    }
+
+    // then we define a custom module
+
+    pub struct FooModule;
+
+    impl Module for FooModule {
+        type Context = ();
+        type GasConfig = FooGasConfig;
+    }
+
+    // we can charge gas using our custom unit to define the price
+
+    impl FooModule {
+        pub fn some_cool_function(ws: &mut WorkingSet<()>) -> Result<()> {
+            Self::charge_gas(
+                ws,
+                "some_other_operation",
+                &String::from("price can be anything"),
+            )
+        }
+    }
+}
+
+#[test]
+fn it_works() {
+    use foo::FooModule;
+    use sov_modules_api::WorkingSet;
+
+    let gas = 1_000_000;
```

Comment:
here we crate a GasoMeter with `gas_price, remining_funds` and pass it to WS. In the RPC queries we can set the GasoMeter to ignore gas counting.